atom/browser/ui/views/autofill_popup_view.cc

```diff
@@ -0,0 +1,469 @@
+#include "ui/gfx/geometry/point.h"
+#include "ui/gfx/geometry/rect.h"
+#include "ui/gfx/text_utils.h"
+#include "ui/views/border.h"
+#include "ui/views/focus/focus_manager.h"
+#include "ui/views/widget/widget.h"
+
+namespace atom {
+
+AutofillPopupView::AutofillPopupView(
+    AutofillPopup* popup,
+    views::Widget* parent_widget)
+    : popup_(popup),
+      parent_widget_(parent_widget),
+      view_proxy_(nullptr),
+      weak_ptr_factory_(this) {
+  CreateChildViews();
+  SetFocusBehavior(FocusBehavior::ALWAYS);
+  set_drag_controller(this);
+}
+
+AutofillPopupView::~AutofillPopupView() {
+  if (popup_) {
+    auto host = popup_->web_contents_->GetRenderViewHost()->GetWidget();
+    host->RemoveKeyPressEventCallback(keypress_callback_);
+    popup_ = nullptr;
+  }
+
+  RemoveObserver();
+
+  if (view_proxy_.get()) {
+    view_proxy_->ResetView();
+  }
+
+  if (GetWidget()) {
+    GetWidget()->Close();
+  }
+}
+
+void AutofillPopupView::Show() {
+  if (!popup_)
+    return;
+
+  const bool initialize_widget = !GetWidget();
+  if (initialize_widget) {
+    parent_widget_->AddObserver(this);
+    views::FocusManager* focus_manager = parent_widget_->GetFocusManager();
+    focus_manager->RegisterAccelerator(
+        ui::Accelerator(ui::VKEY_RETURN, ui::EF_NONE),
+        ui::AcceleratorManager::kNormalPriority,
+        this);
+    focus_manager->RegisterAccelerator(
+        ui::Accelerator(ui::VKEY_ESCAPE, ui::EF_NONE),
+        ui::AcceleratorManager::kNormalPriority,
+        this);
+
+    // The widget is destroyed by the corresponding NativeWidget, so we use
+    // a weak pointer to hold the reference and don't have to worry about
+    // deletion.
+    views::Widget* widget = new views::Widget;
+    views::Widget::InitParams params(views::Widget::InitParams::TYPE_POPUP);
+    params.delegate = this;
+    params.parent = parent_widget_->GetNativeView();
+    widget->Init(params);
+
+    // No animation for popup appearance (too distracting).
+    widget->SetVisibilityAnimationTransition(views::Widget::ANIMATE_HIDE);
+
+    show_time_ = base::Time::Now();
+  }
+
+  SetBorder(views::CreateSolidBorder(
+      kPopupBorderThickness,
+      GetNativeTheme()->GetSystemColor(
+          ui::NativeTheme::kColorId_UnfocusedBorderColor)));
+
+  DoUpdateBoundsAndRedrawPopup();
+  GetWidget()->Show();
+
+  if (initialize_widget)
+    views::WidgetFocusManager::GetInstance()->AddFocusChangeListener(this);
+
+  keypress_callback_ = base::Bind(&AutofillPopupView::HandleKeyPressEvent,
+    base::Unretained(this));
+  auto host = popup_->web_contents_->GetRenderViewHost()->GetWidget();
+  host->AddKeyPressEventCallback(keypress_callback_);
+}
+
+void AutofillPopupView::Hide() {
+  if (popup_) {
+    auto host = popup_->web_contents_->GetRenderViewHost()->GetWidget();
+    host->RemoveKeyPressEventCallback(keypress_callback_);
+    popup_ = NULL;
+  }
+
+  RemoveObserver();
+
+  if (GetWidget()) {
+    GetWidget()->Close();
+  } else {
+    delete this;
```

Comment:
I have removed that `delete this` call, but since the view is also `WidgetDelegate`, it gets deleted when the `Widget` gets deleted, so we can't manage the lifetime ourselves. (It would crash if I tried)